File: android/src/main/java/com/vineyardbovines/nitro/stylesheet/HybridNitroDeviceModule.kt
package "com.vineyardbovines.nitro.stylesheet"

import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.AnyMap
import com.margelo.nitro.core.AnyValue
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import android.accessibilityservice.AccessibilityServiceInfo
import android.content.Context
import android.provider.Settings
import android.view.accessibility.AccessibilityManager
import androidx.core.content.ContextCompat
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

@Keep
@DoNotStrip
class HybridNitroDeviceModule: HybridNitroDeviceModuleSpec() {
    private val TAG = "HybridNitroDeviceModule"

    override val memorySize: Long
        get() = 0L

    override fun getDeviceTraits(): DeviceTraits = DeviceTraits(
        colorScheme = getColorScheme(),
        screenDimensions = getScreenDimensions(),
        contentSizeCategory = getContentSizeCategory(),
        screenInsets = getScreenInsets(),
        displayScale = getDisplayScale(),
        fontScale = getFontScale(),
        isRTL = getIsRTL(),
        screenOrientation = getScreenOrientation(),
        pixelRatio = getPixelRatio()
    )

    override fun runHaptic(type: HapticTypes) {
        val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibratorManager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            vibratorManager.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }

        if (!vibrator.hasVibrator()) return

        when (type) {
            HapticTypes.SUCCESS, HapticTypes.WARNING, HapticTypes.ERROR -> {
                val duration = when (type) {
                    HapticTypes.SUCCESS -> 50L
                    HapticTypes.WARNING -> 100L
                    else -> 150L
                }
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    vibrator.vibrate(VibrationEffect.createOneShot(duration, VibrationEffect.DEFAULT_AMPLITUDE))
                } else {
                    @Suppress("DEPRECATION")
                    vibrator.vibrate(duration)
                }
            }
            HapticTypes.LIGHT -> performHapticFeedback(20)
            HapticTypes.MEDIUM -> performHapticFeedback(40)
            HapticTypes.HEAVY -> performHapticFeedback(60)
            HapticTypes.SOFT -> performHapticFeedback(30)
            HapticTypes.RIGID -> performHapticFeedback(50)
            HapticTypes.SELECT -> performHapticFeedback(10)
        }
    }

    private fun performHapticFeedback(duration: Long) {
        val vibrator = context.getSystemService<Vibrator>() ?: return
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(duration, VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            @Suppress("DEPRECATION")
            vibrator.vibrate(duration)
        }
    }

    override fun getColorScheme(): ColorScheme = when (
        context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
    ) {
        Configuration.UI_MODE_NIGHT_YES -> ColorScheme.DARK
        Configuration.UI_MODE_NIGHT_NO -> ColorScheme.LIGHT
        else -> ColorScheme.UNSPECIFIED
    }

    override fun getScreenDimensions(): ScreenDimensions {
        val metrics = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            context.getSystemService<DisplayManager>()?.displays?.firstOrNull()?.let { display ->
                DisplayMetrics().also { display.getRealMetrics(it) }
            } ?: context.resources.displayMetrics
        } else {
            @Suppress("DEPRECATION")
            context.resources.displayMetrics
        }

        return ScreenDimensions(
            width = metrics.widthPixels.toDouble(),
            height = metrics.heightPixels.toDouble()
        )
    }

    override fun getContentSizeCategory(): String {
        val fontScale = context.resources.configuration.fontScale
        return when {
            fontScale <= 0.85f -> "xSmall"
            fontScale <= 0.95f -> "Small"
            fontScale <= 1.05f -> "Medium"
            fontScale <= 1.15f -> "Large"
            fontScale <= 1.25f -> "xLarge"
            fontScale <= 1.35f -> "xxLarge"
            fontScale <= 1.45f -> "xxxLarge"
            fontScale <= 1.55f -> "accessibilityMedium"
            fontScale <= 1.65f -> "accessibilityLarge"
            fontScale <= 1.75f -> "accessibilityExtraLarge"
            fontScale <= 1.85f -> "accessibilityExtraExtraLarge"
            else -> "accessibilityExtraExtraExtraLarge"
        }
    }

    override fun getScreenInsets(): ScreenInsets {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val metrics = context.getSystemService<DisplayManager>()?.displays?.firstOrNull()?.let { display ->
                context.getSystemService(Context.WINDOW_SERVICE)?.let { windowManager ->
                    windowManager::class.java.getMethod("getCurrentWindowMetrics").invoke(windowManager)
                }
            }

            val insets = (metrics?.javaClass?.getMethod("getWindowInsets")?.invoke(metrics) as? WindowInsets)
                ?.getInsets(WindowInsets.Type.systemBars())

            if (insets != null) {
                return ScreenInsets(
                    top = insets.top.toDouble(),
                    left = insets.left.toDouble(),
                    bottom = insets.bottom.toDouble(),
                    right = insets.right.toDouble()
                )
            }
        }
        return ScreenInsets(0.0, 0.0, 0.0, 0.0)
    }

    override fun getDisplayScale(): Double =
        context.resources.displayMetrics.density.toDouble()

    override fun getFontScale(): Double =
        context.resources.configuration.fontScale.toDouble()

    override fun getIsRTL(): Boolean =
        context.resources.configuration.layoutDirection == Configuration.LAYOUTDIR_RTL

    override fun getScreenOrientation(): ScreenOrientation {
        val dimensions = getScreenDimensions()
        return if (dimensions.width > dimensions.height) {
            ScreenOrientation.LANDSCAPE
        } else {
            ScreenOrientation.PORTRAIT
        }
    }

    override fun getPixelRatio(): Double =
        context.resources.displayMetrics.density.toDouble()
}

File: ios/HybridNitroDeviceModule.swift
import Foundation
import NitroModules
import Combine
import UIKit

class HybridNitroDeviceModule : HybridNitroDeviceModuleSpec {
    private var traitsListener: ((Any) -> Void)?
    private var cancellable: AnyCancellable?

    public var hybridContext = margelo.nitro.HybridContext()

    public var memorySize: Int {
        return getSizeOf(self)
    }

    public func getDeviceTraits() throws -> DeviceTraits {
        let colorScheme = try self.getColorScheme()
        let screenDimensions = try self.getScreenDimensions()
        let contentSizeCategory = self.getContentSizeCategory()
        let screenInsets = try self.getScreenInsets()
        let displayScale = try self.getDisplayScale()
        let fontScale = try self.getFontScale()
        let isRTL = try self.getIsRTL()
        let screenOrientation = self.getScreenOrientation()
        let pixelRatio = self.getPixelRatio()

        return DeviceTraits(
            colorScheme: colorScheme,
            screenDimensions: screenDimensions,
            contentSizeCategory: contentSizeCategory,
            screenInsets: screenInsets,
            displayScale: displayScale,
            fontScale: fontScale,
            isRTL: isRTL,
            screenOrientation: screenOrientation,
            pixelRatio: pixelRatio
        )
    }

    public func runHaptic(type: HapticTypes) throws {
        switch type {
        case .success, .warning, .error:
            let generator = UINotificationFeedbackGenerator()
            generator.prepare()
            generator.notificationOccurred(toNotificationFeedbackType(type: type))

        case .light, .medium, .heavy, .soft, .rigid:
            let generator = UIImpactFeedbackGenerator(style: toImpactFeedbackStyle(style: type))
            generator.prepare()
            generator.impactOccurred()

        case .select:
            let generator = UISelectionFeedbackGenerator()
            generator.prepare()
            generator.selectionChanged()
        }
    }

    private func toNotificationFeedbackType(type: HapticTypes) -> UINotificationFeedbackGenerator.FeedbackType {
        switch type {
        case .success: return .success
        case .warning: return .warning
        case .error: return .error
        default: return .success
        }
    }

    private func toImpactFeedbackStyle(style: HapticTypes) -> UIImpactFeedbackGenerator.FeedbackStyle {
        switch style {
        case .light: return .light
        case .medium: return .medium
        case .heavy: return .heavy
        case .soft: return .soft
        case .rigid: return .rigid
        default: return .medium
        }
    }

    public func getColorScheme() throws -> ColorScheme {
        let interfaceStyle = UIScreen.main.traitCollection.userInterfaceStyle

        switch interfaceStyle {
        case .dark:
            return ColorScheme.dark
        case .light:
            return ColorScheme.light
        case .unspecified:
            return ColorScheme.unspecified
        default:
            return ColorScheme.unspecified
        }
    }

    public func getDisplayScale() throws -> Double {
        func getDisplayScaleFn() -> Double {
            guard let window = self.getWindow() else {
                return 1
            }

            return window.screen.scale
        }

        if Thread.isMainThread {
            return getDisplayScaleFn()
        }

        return DispatchQueue.main.sync {
            return getDisplayScaleFn()
        }
    }

    public func getFontScale() throws -> Double {
        func getFontScaleFn() -> Double {
            let contentSizeCategory = UIApplication.shared.preferredContentSizeCategory
            let defaultMultiplier = 17.0

            switch contentSizeCategory {
            case .extraExtraExtraLarge:
                return 23.0 / defaultMultiplier
            case .extraExtraLarge:
                return 21.0 / defaultMultiplier
            case .extraLarge:
                return 19.0 / defaultMultiplier
            case .large:
                return 17.0 / defaultMultiplier
            case .medium:
                return 16.0 / defaultMultiplier
            case .small:
                return 15.0 / defaultMultiplier
            case .extraSmall:
                return 14.0 / defaultMultiplier
            case .accessibilityMedium:
                return 29.0 / defaultMultiplier
            case .accessibilityLarge:
                return 33.0 / defaultMultiplier
            case .accessibilityExtraLarge:
                return 40.0 / defaultMultiplier
            case .accessibilityExtraExtraLarge:
                return 47.0 / defaultMultiplier
            case .accessibilityExtraExtraExtraLarge:
                return 53.0 / defaultMultiplier
            default:
                return 1.0
            }
        }

        if Thread.isMainThread {
            return getFontScaleFn()
        }

        return DispatchQueue.main.sync {
            return getFontScaleFn()
        }
    }

    private func getScreenDimensionsFn() -> ScreenDimensions {
        let window = self.getWindow()

        guard let windowFrame = window?.frame else {
            return ScreenDimensions(width: 0, height: 0)
        }

        return ScreenDimensions(
            width: windowFrame.size.width,
            height: windowFrame.size.height
        )
    }

    public func getScreenDimensions() throws -> ScreenDimensions {
        if Thread.isMainThread {
            return self.getScreenDimensionsFn()
        }

        return DispatchQueue.main.sync {
            return self.getScreenDimensionsFn()
        }
    }

    public func getScreenOrientation() -> ScreenOrientation {
        let screenDimensions = self.getScreenDimensionsFn()

        if (screenDimensions.width > screenDimensions.height) {
            return ScreenOrientation.landscape
        }

        return ScreenOrientation.portrait
    }

    public func getContentSizeCategory() -> String {
        func getContentSizeCategoryFn() -> String {
            let contentSizeCategory = UIApplication.shared.preferredContentSizeCategory

            switch contentSizeCategory {
            case .extraExtraExtraLarge:
                return "xxxLarge"
            case .extraExtraLarge:
                return "xxLarge"
            case .extraLarge:
                return "xLarge"
            case .large:
                return "Large"
            case .medium:
                return "Medium"
            case .small:
                return "Small"
            case .extraSmall:
                return "xSmall"
            case .accessibilityMedium:
                return "accessibilityMedium"
            case .accessibilityLarge:
                return "accessibilityLarge"
            case .accessibilityExtraLarge:
                return "accessibilityExtraLarge"
            case .accessibilityExtraExtraLarge:
                return "accessibilityExtraExtraLarge"
            case .accessibilityExtraExtraExtraLarge:
                return "accessibilityExtraExtraExtraLarge"
            default:
                return "unspecified"
            }
        }

        if Thread.isMainThread {
            return getContentSizeCategoryFn()
        }

        return DispatchQueue.main.sync {
            return getContentSizeCategoryFn()
        }
    }

    public func getScreenInsets() throws -> ScreenInsets {
        func getInsetsFn() -> ScreenInsets {
            let window = self.getWindow()

            guard let insets = window?.safeAreaInsets else {
                return ScreenInsets(top: 0, left: 0, bottom: 0, right: 0)
            }

            return ScreenInsets(
                top: insets.top,
                left: insets.left,
                bottom: insets.bottom,
                right: insets.right
            )
        }

        if Thread.isMainThread {
            return getInsetsFn()
        }

        return DispatchQueue.main.sync {
            return getInsetsFn()
        }
    }

    public func getIsRTL() throws -> Bool {
        func getIsRTLFn() -> Bool {
            let hasForcedRtl = UserDefaults.standard.bool(forKey: "RCTI18nUtil_forceRTL")
            let isRtl = UIApplication.shared.userInterfaceLayoutDirection == .rightToLeft

            return hasForcedRtl || isRtl
        }

        if Thread.isMainThread {
            return getIsRTLFn()
        }

        return DispatchQueue.main.sync {
            return getIsRTLFn()
        }
    }

    public func getPixelRatio() -> Double {
        func getPixelRatioFn() -> Double {
            let window = self.getWindow()

            guard let window else {
                return 1;
            }

            return window.screen.scale
        }

        if Thread.isMainThread {
            return getPixelRatioFn()
        }

        return DispatchQueue.main.sync {
            return getPixelRatioFn()
        }
    }

    private func getWindow() -> UIWindow? {
        if #available(iOS 15.0, *) {
            guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                let window = windowScene.windows.first else {
                return nil
            }
            return window
        } else {
            return UIApplication.shared.windows.first
        }
    }
}
